src/oracle/middleware.js

```diff
@@ -0,0 +1,761 @@
+
+    //check if time for update
+    query = await queryTable("rounds", ["started_at"], jobName)
+
+    let timeForUpdate = (Date.now()/1000) >= query.started_at + Number(pushInterval)
+
+    //if there is no last price, if it is time for a price update or if there is a new round, update price
+    let toUpdate = lastPrice == -1 || lastPrice == 0 || requestType == 1 && timeForUpdate || requestType == 3
+    //if last price is found and it is a price deviation request
+    if (lastPrice != -1 && requestType == 2) {
+      
+      //get decimal places for feed
+      let decimalPlaces = feeds[jobName].decimalPlaces
+      //calculate percentage change
+      lastPrice = lastPrice * Math.pow(10, Number(decimalPlaces))
+
+      let percChange = Math.abs((result - lastPrice) / lastPrice) * 100
+      console.log("Price change is " + percChange + "%. Last Price: " + String(result) + ". Current Price: " + String(lastPrice))
+
+      //get price deviation threshold for feed
+      let priceDeviationPercentage = feeds[jobName].priceDeviationPerc
+      //update price if result is greater than price deviation threshold
+      toUpdate = percChange > priceDeviationPercentage
+    }
+
+    //get seconds since last price submission
+    query = await queryTable("jobs", ["last_submission_time"], jobName)
+    let timePassedSinceSubmission = (Date.now()/1000) - query.last_submission_time
+    //check if in submission
+    let inSubmission = timePassedSinceSubmission < Number(SEND_CHECK_INTERVAL)
+
+    /**
+      * If an update needs to happen
+      * An update happens for the following reasons
+      *    - First request
+      *    - Job request was because time expired
+      *    - Price deviation found
+      *    - PLUS not already waiting for a submission
+      */
+    if (toUpdate && !inSubmission) {
+
+      //get latest round
+      let latestRound = await queryRound(jobName)
+      await updateTable("rounds", latestRound, jobName)
+
+      //get the round for submission
+      let query = await queryTable("jobs", ["last_reported_round"], jobName)
+      let lastReportedRound = query.last_reported_round
+      let lastRoundId = isNaN(latestRound.round_id) ? lastReportedRound : latestRound.round_id
+      let roundToSubmit = lastReportedRound < lastRoundId ? lastRoundId : lastRoundId + 1
+
+      //check if new round
+      let newRound = roundToSubmit != lastRoundId
+
+      //push price on chain if first round, haven't started previous round and have not submitted yet in the same round
+      if (roundToSubmit == 1 || (newRound && latestRound.started_by != FROM) || (!newRound && !latestRound.submission_made)) {
+        console.log("Updating price for round", roundToSubmit)
+
+        let submitted = await pushPrice(result, jobName, roundToSubmit, FROM)
+
+        //update last reported round
+        if (submitted){
+          await updateTable("jobs", {"last_reported_round": roundToSubmit}, jobName)
```

Comment:
IDK if this is critical cause IDK how often this will be called but usually if you have multiple updates in the same call you batch them together to send a single transaction to the DB

Reply:
Since there are calls and conditions in between, it is fine as it is in this case because we want the state updated as soon as possible without waiting for any external checks JIC the state is queried from some other part of the middleware at the same time